cmd/snap-seccomp/main.go

```diff
@@ -619,21 +623,39 @@ func compile(content []byte, out string) error {
 	for scanner.Scan() {
 		line := strings.TrimSpace(scanner.Text())
 
-		// FIXME: right now complain mode is the equivalent to
-		// unrestricted.  We'll want to change this once we
-		// seccomp logging is in order.
-		//
 		// special case: unrestricted means we switch to an allow-all
 		//               filter and are done
-		if line == "@unrestricted" || line == "@complain" {
+		if line == "@unrestricted" {
 			secFilter, err = seccomp.NewFilter(seccomp.ActAllow)
 			if err != nil {
-				return fmt.Errorf("cannot create seccomp filter: %s", err)
+				return fmt.Errorf("cannot create unrestricted seccomp filter: %s", err)
 			}
 			if err := addSecondaryArches(secFilter); err != nil {
 				return err
 			}
 			break
+		} else if line == "@complain" {
+			var fallback bool = false
+
+			secFilter, err = seccomp.NewFilter(seccomp.ActLog)
+			if err != nil {
+				// ActLog is only supported in newer versions
+				// of the kernel, libseccomp, and
+				// libseccomp-golang. Attempt to fall back to
+				// ActAllow before erroring out.
+				fallback = true
+				secFilter, err = seccomp.NewFilter(seccomp.ActAllow)
+				if err != nil {
+					return fmt.Errorf("cannot create complain seccomp filter: %s", err)
+				}
+			}
+			if err := addSecondaryArches(secFilter); err != nil {
+				return err
+			}
+
+			if fallback {
```

Comment:
Can you add a comment here stating that this fallback is resorting to the old pre-logging behavior of just allowing everything?

Reply:
Yes, good idea.